include/pybind11/cast.h

```diff
@@ -114,6 +162,46 @@ PYBIND11_NOINLINE inline handle get_type_handle(const std::type_info &tp, bool t
     return handle(type_info ? ((PyObject *) type_info->type) : nullptr);
 }
 
+struct value_and_holder {
+    void *&value, *&holder;
+    value_and_holder(void *&v, void *&h) : value{v}, holder{h} {}
+    template <typename V> enable_if_t<std::is_pointer<V>::value, V> value_as() const { return reinterpret_cast<V>(value); }
+    template <typename H> enable_if_t<std::is_pointer<H>::value, H> holder_as() const { return reinterpret_cast<H>(holder); }
```

Comment:
(Replaced it entirely: it is now renamed `value_ptr<V>` and returns a `V*` instead; the holder version similarly returns a reference instead of a pointer.)